### PR TITLE
#157: Added new capabilities for LISTAGG API.

### LINKED ISSUE #157: New LISTAGG API
Up to now `LISTAGG` is pushed down as `GROUP_CONCAT`. We want to change that to include the `LISTAGG` feature `ON OVERFLOW TRUNCATE` which is relevant if the result exceeds the 2000000 character limit. Besides that, the API is very similar to `GROUP_CONCAT`.

### Example Queries

```sql
SELECT LISTAGG(a ON OVERFLOW ERROR) FROM t;
SELECT LISTAGG(DISTINCT a ON OVERFLOW ERROR) FROM t;
SELECT LISTAGG(a, ', ' ON OVERFLOW ERROR) FROM t;
SELECT LISTAGG(a ON OVERFLOW TRUNCATE WITH COUNT) FROM t;
SELECT LISTAGG(a ON OVERFLOW TRUNCATE WITHOUT COUNT) FROM t;
SELECT LISTAGG(a ON OVERFLOW TRUNCATE 'filler' WITH COUNT) FROM t;
SELECT LISTAGG(a ON OVERFLOW TRUNCATE 'filler' WITHOUT COUNT) FROM t; 
SELECT LISTAGG(a ON OVERFLOW ERROR) WITHIN GROUP (ORDER BY b) FROM t;
SELECT LISTAGG(DISTINCT a, ', ' ON OVERFLOW TRUNCATE 'filler' WITH COUNT) WITHIN GROUP (ORDER BY b DESC NULLS LAST) FROM t;
```

### New Capabilities

* `FN_AGG_LISTAGG`
* `FN_AGG_LISTAGG_DISTINCT`
* `FN_AGG_LISTAGG_SEPARATOR`
* `FN_AGG_LISTAGG_ON_OVERFLOW_ERROR`
* `FN_AGG_LISTAGG_ON_OVERFLOW_TRUNCATE`
* `FN_AGG_LISTAGG_ORDER_BY`

### Aggregate Function

New special case for aggregate functions:

`LISTAGG([DISTINCT] arg1[, separator] ON OVERFLOW {ERROR | TRUNCATE [truncationFiller] {WITH | WITHOUT} COUNT}) [WITHIN GROUP (orderBy)]` (requires set-function capability `LISTAGG`)

```json
{
    "type": "function_aggregate_listagg",
    "name": "LISTAGG",
    "distinct": true,
    "arguments": [
    {
        ...
    }
    ],
    "separator": ", ",
    "overflowBehavior":
    {
        "type": "TRUNCATE",
        "truncationType": "WITH COUNT",
        "truncationFiller": "..."
    },
    "orderBy": [
        {
            "type": "order_by_element",
            "expression":
            {
              "type": "column",
               "columnNr": 1,
                "name": "USER_ID",
                "tableName": "CLICKS"
            },
            "isAscending": true,
            "nullsLast": true
        }
    ]
}
```

Notes:
* `distinct`: Optional. Requires set-function capability `LISTAGG_DISTINCT`.
* `separator`: Optional. Requires set-function capability `LISTAGG_SEPARATOR`.
* `overflowBehavior`: `type` is `"ERROR"` (requires set-function capability `LISTAGG_ON_OVERFLOW_ERROR`) or `"TRUNCATE"` (requires set-function capability `LISTAGG_ON_OVERFLOW_TRUNCATE`). Only for `"TRUNCATE"` the members `truncationType` and optionally `truncationFiller` exist. `truncationType` is `"WITH COUNT"` or `"WITHOUT COUNT"`.
* `orderBy`: Optional. The requested order-by clause, a list of `order_by_element` elements. The field `expression` contains the expression to order by. The `group by` clause of a `SELECT` query uses the same `order_by_element` element type. The clause requires the set-function capability `LISTAGG_ORDER_BY`.

### PR DESCRIPTION
**Pull request by AnastasiiaSergienko**
Closes #157

### CODE REVIEW

File: doc/development/api/capabilities_list.md [hunk 1]
```diff
@@ -261,6 +261,12 @@ Here you can find a list of available Virtual Schema capabilities.
 - FN_AGG_GROUP_CONCAT_ORDER_BY
 - FN_AGG_GROUP_CONCAT_SEPARATOR
 - FN_AGG_LAST_VALUE
+- FN_AGG_LISTAGG
+- FN_AGG_LISTAGG_DISTINCT
+- FN_AGG_LISTAGG_SEPARATOR
+- FN_AGG_LISTAGG_ON_OVERFLOW_ERROR
+- FN_AGG_LISTAGG_ON_OVERFLOW_TRUNCATE
+- FN_AGG_LISTAGG_ORDER_BY
```
Comment (exanm) [inline doc/development/api/capabilities_list.md:269]:
```suggestion
- FN_AGG_LISTAGG_ON_OVERFLOW_ERROR
- FN_AGG_LISTAGG_ON_OVERFLOW_TRUNCATE
- FN_AGG_LISTAGG_ORDER_BY
- FN_AGG_LISTAGG_SEPARATOR
```

File: doc/development/api/virtual_schema_api.md [hunk 2]
```diff
@@ -1214,3 +1214,49 @@ Notes:
 * `distinct`: Optional. Requires set-function capability `GROUP_CONCAT_DISTINCT.`
 * `orderBy`: Optional. The requested order-by clause, a list of `order_by_element` elements. The field `expression` contains the expression to order by. The `group by` clause of a `SELECT` query uses the same `order_by_element` element type. The clause requires the set-function capability `GROUP_CONCAT_ORDER_BY`.
 * `separator`: Optional. Requires set-function capability `GROUP_CONCAT_SEPARATOR`.
+
+##### LISTAGG
+
+`LISTAGG([DISTINCT] arg1[, separator] ON OVERFLOW {ERROR | TRUNCATE [truncationFiller] {WITH | WITHOUT} COUNT}) [WITHIN GROUP (orderBy)]`
+ (requires set-function capability LISTAGG)
```
Comment (exanm) [inline doc/development/api/virtual_schema_api.md:1221]:
```suggestion
 (requires set-function capability `LISTAGG`)
```

File: doc/development/api/virtual_schema_api.md [hunk 3]
```diff
@@ -1214,3 +1214,49 @@ Notes:
...
+    ]
+}
+```
+
+Notes:
+
+* `distinct`: Optional. Requires set-function capability `LISTAGG_DISTINCT`.
+* `separator`: Optional. Requires set-function capability `LISTAGG_SEPARATOR`.
+* `overflowBehavior`: type is `ERROR` (requires set-function capability `LISTAGG_ON_OVERFLOW_ERROR`) or `TRUNCATE` (requires set-function capability `LISTAGG_ON_OVERFLOW_TRUNCATE`). Only for `TRUNCATE` the members `truncationType` and optionally `truncationFiller` exist. `truncationType` is `WITH COUNT` or `WITHOUT COUNT`.
+* `orderBy`: Optional. The requested order-by clause, a list of order_by_element elements. The field expression contains the expression to order by. The group by clause of a `SELECT` query uses the same order_by_element element type. The clause requires the set-function capability `LISTAGG_ORDER_BY`.
```
Comment (exanm) [inline doc/development/api/virtual_schema_api.md:1262]: Copied the formatting from GROUP_CONCAT:
```suggestion
* `orderBy`: Optional. The requested order-by clause, a list of `order_by_element` elements. The field `expression` contains the expression to order by. The `group by` clause of a `SELECT` query uses the same `order_by_element` element type. The clause requires the set-function capability `LISTAGG_ORDER_BY`.
```

File: doc/development/api/virtual_schema_api.md [hunk 4]
```diff
@@ -1214,3 +1214,49 @@ Notes:
...
+        }
+    ]
+}
+```
+
+Notes:
+
+* `distinct`: Optional. Requires set-function capability `LISTAGG_DISTINCT`.
+* `separator`: Optional. Requires set-function capability `LISTAGG_SEPARATOR`.
+* `overflowBehavior`: type is `ERROR` (requires set-function capability `LISTAGG_ON_OVERFLOW_ERROR`) or `TRUNCATE` (requires set-function capability `LISTAGG_ON_OVERFLOW_TRUNCATE`). Only for `TRUNCATE` the members `truncationType` and optionally `truncationFiller` exist. `truncationType` is `WITH COUNT` or `WITHOUT COUNT`.
```
Comment (exanm) [inline doc/development/api/virtual_schema_api.md:1261]: `type`:
```suggestion
* `overflowBehavior`: `type` is `ERROR` (requires set-function capability `LISTAGG_ON_OVERFLOW_ERROR`) or `TRUNCATE` (requires set-function capability `LISTAGG_ON_OVERFLOW_TRUNCATE`). Only for `TRUNCATE` the members `truncationType` and optionally `truncationFiller` exist. `truncationType` is `WITH COUNT` or `WITHOUT COUNT`.
```
Also, for IS JSON and JSON_VALUE we included the double quotes like this:
```suggestion
* `overflowBehavior`: `type` is `"ERROR"` (requires set-function capability `LISTAGG_ON_OVERFLOW_ERROR`) or `"TRUNCATE"` (requires set-function capability `LISTAGG_ON_OVERFLOW_TRUNCATE`). Only for `"TRUNCATE"` the members `truncationType` and optionally `truncationFiller` exist. `truncationType` is `"WITH COUNT"` or `"WITHOUT COUNT"`.
```
Should we do the same here? Or adapt the JSON description?

**Review by exanm — DISMISSED** [review]
See suggested changes.

File: doc/development/api/virtual_schema_api.md (hunk 4 above)
Comment (AnastasiiaSergienko) [reply doc/development/api/virtual_schema_api.md]: It's okay to make it similar to the JSON. We can change the formatting later

**Pull request merged**